graphhopper.sh

```diff
@@ -2,68 +2,97 @@
+  echo "     --action measurement does performance analysis of the current source version via random routes (Measurement class)"
+  echo "     --action torture     can be used to test real world routes via feeding graphhopper logs into a GraphHopper system (Torture class)"
+  echo "-d | --datareader.file: path to the map file or name of the file to download"
+  echo "-o | -gc | --graph-cache: directory for graph cache output"
+  echo "-c | --config: app configuration"
+  echo "-fd| --force-download: force the download of the data file if needed"
+  echo "-rb| --run-background: run app in background"
+  echo "-jar: changes the jar file"
+}
 
-ACTION=$1
-FILE=$2
+while [ ! -z $1 ]; do
+  case $1 in
+    -h|--help) printBashUsage
+      exit 0;;
+    -fd|--force-download) FORCE_DWN=1; shift 1;;
+    clean) ACTION=clean; shift 1;;
+    -a|--action) ACTION=$2; shift 2;;
+    -f|--datareader.file) FILE="$2"; shift 2;;
+    -c|--config) CONFIG="$2"; shift 2;;
+    -rb|--run-background) RUN_BACKGROUND=$2; shift 2;;
+    -o|-gc|--graph-cache) GRAPH="$2"; shift 2;;
+    -jar) JAR="$2"; shift 2;;
+    -*|--*) echo "Option unknown: $1"
+        echo
+        printBashUsage
+	exit 2;;
+    # backward compatibility
+    *) REMAINING_ARGS+=($1); shift 1;;
+  esac
+done
+
+if [ ! -z $REMAINING_ARGS ]; then
+  echo "Remaining arguments: ${REMAINING_ARGS[@]}"
+fi
 
-function printUsage {
- echo
- echo "./graphhopper.sh import|web <some-map-file>"
- echo "./graphhopper.sh clean|build|buildweb|help"
- echo
- echo "  help        this message"
- echo "  import      creates the graphhopper files used for later (faster) starts"
- echo "  web         starts a local server for user access at localhost:8989 and API access at localhost:8989/route"
- echo "  build       creates the graphhopper JAR (without the web module)"
- echo "  buildweb    creates the graphhopper JAR (with the web module)"
- echo "  clean       removes all JARs, necessary if you need to use the latest source (e.g. after switching the branch etc)"
- echo "  measurement does performance analysis of the current source version via random routes (Measurement class)"
- echo "  torture     can be used to test real world routes via feeding graphhopper logs into a GraphHopper system (Torture class)"
- echo "  miniui      is a simple Java/Swing visualization application used for debugging purposes (MiniGraphUI class)"
- echo "  extract     calls the overpass API to grab any area as .osm file"
- echo "  android     builds, deploys and starts the Android demo for your connected device"
-}
+if [ -z $ACTION ] || [ -z $FILE ]; then
+  ACTION=${REMAINING_ARGS[0]}
+  FILE=${REMAINING_ARGS[1]}
+fi
 
 if [ "$ACTION" = "" ]; then
- echo "## action $ACTION not found. try" 
- printUsage
+ echo "## action $ACTION not found!"
+ printBashUsage
+fi
+
+# default init, https://stackoverflow.com/a/28085062/194609
+: "${CONFIG:=config.yml}"
```

Comment:
`:` is a easy-to-miss synonym for `true`. This form is clearer:

    CONFIG=${CONFIG:-config.yml}

and doesn't require studying a stackoverflow post.

Reply:
Nice update.
I was reading again the flags, and I have perhaps others suggestions:
- -r for --datareader.file
- -d for --run-backgound or --detach (which is really common for different software)
- -f for --force-download (it can still be -fd!)

Reply:
@legraina Thanks! For `-r` I'm not sure. I think -f for "file" is not uncommon. Maybe -i or -in for input as we have -o for output ...

> and doesn't require studying a stackoverflow post.

@justinmk2 the other method requires typing the variable twice ;) (maybe useful later for long variables) and to be honest the `:-` is not 100% clear too and the `:=` smells more intutive